// @akashic-extension/akashic-animation/lib/AttrId.js
"use strict";
/**
 * 属性ID
 *
 * Posture#attrs 属性配列に格納された値にアクセスするための添字として用いる。
 * 例えばX方向移動量にアクセスするには posture.attrs[AttrId.tx] とする。
 */
var AttrId;
(function (AttrId) {
    /**
     * X方向移動。
     */
    AttrId[AttrId["tx"] = 0] = "tx";
    /**
     * Y方向移動。
     */
    AttrId[AttrId["ty"] = 1] = "ty";
    /**
     * Z軸回転(deg)。
     */
    AttrId[AttrId["rz"] = 2] = "rz";
    /**
     * X方向スケール。
     */
    AttrId[AttrId["sx"] = 3] = "sx";
    /**
     * Y方向スケール。
     */
    AttrId[AttrId["sy"] = 4] = "sy";
    /**
     * 半透明度。
     *
     * 0で透明、1で不透明を表す。
     */
    AttrId[AttrId["alpha"] = 5] = "alpha";
    /**
     * セル値。
     *
     * CellValueクラスを参照。
     */
    AttrId[AttrId["cv"] = 6] = "cv";
    /**
     * セル回転位置X座標。
     *
     * セルの中心位置は pvtx, pvty 属性によって与えられる。
     */
    AttrId[AttrId["pvtx"] = 7] = "pvtx";
    /**
     * セル回転位置Y座標。
     *
     * セルの中心位置は pvtx, pvty 属性によって与えられる。
     */
    AttrId[AttrId["pvty"] = 8] = "pvty";
    /**
     * セルテクスチャU座標。
     *
     * セルの参照する画像上の参照位置。正規化されている。
     */
    AttrId[AttrId["tu"] = 9] = "tu";
    /**
     * セルテクスチャV座標。
     *
     * セルの参照する画像上の参照位置。正規化されている。
     */
    AttrId[AttrId["tv"] = 10] = "tv";
    /**
     * セル描画優先順位。
     *
     * あるActorをレンダリングするときのセル間の描画優先順位。
     */
    AttrId[AttrId["prio"] = 11] = "prio";
    /**
     * 可視属性。
     */
    AttrId[AttrId["visibility"] = 12] = "visibility";
    /**
     * 円アタリ判定半径。
     */
    AttrId[AttrId["ccr"] = 13] = "ccr";
    /**
     * 水平フリップ。
     */
    AttrId[AttrId["flipH"] = 14] = "flipH";
    /**
     * 垂直フリップ。
     */
    AttrId[AttrId["flipV"] = 15] = "flipV";
    /**
     * ユーザデータ。
     */
    AttrId[AttrId["userData"] = 16] = "userData";
})(AttrId || (AttrId = {}));
module.exports = AttrId;
