// @akashic-extension/akashic-animation/lib/AnimeParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Animation = exports.CurveTie = exports.Curve = exports.KeyFrame = exports.IpCurve = exports.CellValue = void 0;
/**
 * セルアニメーションの値。
 *
 * KeyFrame<CellValue>に格納されセルアニメーション（時間とともに変化するセル）に利用される。
 */
var CellValue = /** @class */ (function () {
    function CellValue() {
    }
    return CellValue;
}());
exports.CellValue = CellValue;
var IpCurve = /** @class */ (function () {
    function IpCurve() {
        this.values = [];
    }
    return IpCurve;
}());
exports.IpCurve = IpCurve;
var KeyFrame = /** @class */ (function () {
    function KeyFrame() {
    }
    return KeyFrame;
}());
exports.KeyFrame = KeyFrame;
var Curve = /** @class */ (function () {
    function Curve() {
        this.attribute = "";
        this.keyFrames = [];
    }
    return Curve;
}());
exports.Curve = Curve;
var CurveTie = /** @class */ (function () {
    function CurveTie() {
        this.boneName = "";
        this.curves = [];
    }
    return CurveTie;
}());
exports.CurveTie = CurveTie;
var Animation = /** @class */ (function () {
    function Animation() {
        this.name = "";
        this.fps = 0;
        this.frameCount = 0;
        this.curveTies = {}; // key = curveTie.boneName
    }
    return Animation;
}());
exports.Animation = Animation;
