// @akashic-extension/akashic-animation/lib/AttrId.js
"use strict";
/**
 * 属性ID
 *
 * Posture#attrs 属性配列に格納された値にアクセスするための添字として用いる。
 * 例えばX方向移動量にアクセスするには posture.attrs[AttrId.tx] とする。
 */
var AttrId;
(function (AttrId) {
    /**
     * X方向移動。
     */
    AttrId[AttrId["tx"] = 0] = "tx";
    /**
     * Y方向移動。
     */
    AttrId[AttrId["ty"] = 1] = "ty";
    /**
     * Z軸回転(deg)。
     */
    AttrId[AttrId["rz"] = 2] = "rz";
    /**
     * X方向スケール。
     */
    AttrId[AttrId["sx"] = 3] = "sx";
    /**
     * Y方向スケール。
     */
    AttrId[AttrId["sy"] = 4] = "sy";
    /**
     * X方向ローカルスケール
     */
    AttrId[AttrId["lsx"] = 5] = "lsx";
    /**
     * Y方向ローカルスケール
     */
    AttrId[AttrId["lsy"] = 6] = "lsy";
    /**
     * 半透明度。
     *
     * 0で透明、1で不透明を表す。
     */
    AttrId[AttrId["alpha"] = 7] = "alpha";
    /**
     * ローカル半透明度。半透明度と違ってこの値は子パーツに継承されない。
     *
     * 0で透明、1で不透明を表す。また、undefinedで未設定を表す。
     */
    AttrId[AttrId["lalpha"] = 8] = "lalpha";
    /**
     * セル値。
     *
     * CellValueクラスを参照。
     */
    AttrId[AttrId["cv"] = 9] = "cv";
    /**
     * セル回転位置X座標。
     *
     * セルの中心位置は pvtx, pvty 属性によって与えられる。
     */
    AttrId[AttrId["pvtx"] = 10] = "pvtx";
    /**
     * セル回転位置Y座標。
     *
     * セルの中心位置は pvtx, pvty 属性によって与えられる。
     */
    AttrId[AttrId["pvty"] = 11] = "pvty";
    /**
     * セルテクスチャU座標。
     *
     * セルの参照する画像上の参照位置。正規化されている。
     */
    AttrId[AttrId["tu"] = 12] = "tu";
    /**
     * セルテクスチャV座標。
     *
     * セルの参照する画像上の参照位置。正規化されている。
     */
    AttrId[AttrId["tv"] = 13] = "tv";
    /**
     * セル描画優先順位。
     *
     * あるActorをレンダリングするときのセル間の描画優先順位。
     */
    AttrId[AttrId["prio"] = 14] = "prio";
    /**
     * イメージの左右反転
     *
     * trueの場合、セル中の画像を左右反転させる
     * セル内での左右反転なので、水平フリップと違ってセルの座標は変わらない
     */
    AttrId[AttrId["iflh"] = 15] = "iflh";
    /**
     * イメージの上下反転
     *
     * trueの場合、セル中の画像を上下反転させる
     * セル内での上下反転なので、垂直フリップと違ってセルの座標は変わらない
     */
    AttrId[AttrId["iflv"] = 16] = "iflv";
    /**
     * 可視属性。
     */
    AttrId[AttrId["visibility"] = 17] = "visibility";
    /**
     * 円アタリ判定半径。
     */
    AttrId[AttrId["ccr"] = 18] = "ccr";
    /**
     * 水平フリップ。
     * trueの場合、セルの原点を支点にしてセルを左右反転させる処理。
     */
    AttrId[AttrId["flipH"] = 19] = "flipH";
    /**
     * 垂直フリップ。
     * trueの場合、セルの原点を支点にして、セルを上下反転させる処理。
     */
    AttrId[AttrId["flipV"] = 20] = "flipV";
    /**
     * ユーザデータ。
     */
    AttrId[AttrId["userData"] = 21] = "userData";
})(AttrId || (AttrId = {}));
module.exports = AttrId;
