// @akashic/playlog/lib/Event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var EventCode;
(function (EventCode) {
    EventCode[EventCode["Join"] = 0] = "Join";
    EventCode[EventCode["Leave"] = 1] = "Leave";
    EventCode[EventCode["Timestamp"] = 2] = "Timestamp";
    EventCode[EventCode["PlayerInfo"] = 3] = "PlayerInfo";
    EventCode[EventCode["Message"] = 32] = "Message";
    EventCode[EventCode["PointDown"] = 33] = "PointDown";
    EventCode[EventCode["PointMove"] = 34] = "PointMove";
    EventCode[EventCode["PointUp"] = 35] = "PointUp";
    EventCode[EventCode["Operation"] = 64] = "Operation";
})(EventCode = exports.EventCode || (exports.EventCode = {}));
var EventIndex;
(function (EventIndex) {
    EventIndex[EventIndex["Code"] = 0] = "Code";
    EventIndex[EventIndex["EventFlags"] = 1] = "EventFlags";
    EventIndex[EventIndex["PlayerId"] = 2] = "PlayerId";
})(EventIndex = exports.EventIndex || (exports.EventIndex = {}));
var EventFlagsMask;
(function (EventFlagsMask) {
    EventFlagsMask[EventFlagsMask["Priority"] = 3] = "Priority";
    EventFlagsMask[EventFlagsMask["Transient"] = 8] = "Transient";
    EventFlagsMask[EventFlagsMask["Ignorable"] = 16] = "Ignorable";
})(EventFlagsMask = exports.EventFlagsMask || (exports.EventFlagsMask = {}));
var JoinEventIndex;
(function (JoinEventIndex) {
    JoinEventIndex[JoinEventIndex["Code"] = 0] = "Code";
    JoinEventIndex[JoinEventIndex["EventFlags"] = 1] = "EventFlags";
    JoinEventIndex[JoinEventIndex["PlayerId"] = 2] = "PlayerId";
    JoinEventIndex[JoinEventIndex["PlayerName"] = 3] = "PlayerName";
    JoinEventIndex[JoinEventIndex["StorageData"] = 4] = "StorageData";
})(JoinEventIndex = exports.JoinEventIndex || (exports.JoinEventIndex = {}));
var LeaveEventIndex;
(function (LeaveEventIndex) {
    LeaveEventIndex[LeaveEventIndex["Code"] = 0] = "Code";
    LeaveEventIndex[LeaveEventIndex["EventFlags"] = 1] = "EventFlags";
    LeaveEventIndex[LeaveEventIndex["PlayerId"] = 2] = "PlayerId";
})(LeaveEventIndex = exports.LeaveEventIndex || (exports.LeaveEventIndex = {}));
var TimestampEventIndex;
(function (TimestampEventIndex) {
    TimestampEventIndex[TimestampEventIndex["Code"] = 0] = "Code";
    TimestampEventIndex[TimestampEventIndex["EventFlags"] = 1] = "EventFlags";
    TimestampEventIndex[TimestampEventIndex["PlayerId"] = 2] = "PlayerId";
    TimestampEventIndex[TimestampEventIndex["Timestamp"] = 3] = "Timestamp";
})(TimestampEventIndex = exports.TimestampEventIndex || (exports.TimestampEventIndex = {}));
var PlayerInfoEventIndex;
(function (PlayerInfoEventIndex) {
    PlayerInfoEventIndex[PlayerInfoEventIndex["Code"] = 0] = "Code";
    PlayerInfoEventIndex[PlayerInfoEventIndex["EventFlags"] = 1] = "EventFlags";
    PlayerInfoEventIndex[PlayerInfoEventIndex["PlayerId"] = 2] = "PlayerId";
    PlayerInfoEventIndex[PlayerInfoEventIndex["PlayerName"] = 3] = "PlayerName";
    PlayerInfoEventIndex[PlayerInfoEventIndex["UserData"] = 4] = "UserData";
})(PlayerInfoEventIndex = exports.PlayerInfoEventIndex || (exports.PlayerInfoEventIndex = {}));
var MessageEventIndex;
(function (MessageEventIndex) {
    MessageEventIndex[MessageEventIndex["Code"] = 0] = "Code";
    MessageEventIndex[MessageEventIndex["EventFlags"] = 1] = "EventFlags";
    MessageEventIndex[MessageEventIndex["PlayerId"] = 2] = "PlayerId";
    MessageEventIndex[MessageEventIndex["Data"] = 3] = "Data";
})(MessageEventIndex = exports.MessageEventIndex || (exports.MessageEventIndex = {}));
var PointDownEventIndex;
(function (PointDownEventIndex) {
    PointDownEventIndex[PointDownEventIndex["Code"] = 0] = "Code";
    PointDownEventIndex[PointDownEventIndex["EventFlags"] = 1] = "EventFlags";
    PointDownEventIndex[PointDownEventIndex["PlayerId"] = 2] = "PlayerId";
    PointDownEventIndex[PointDownEventIndex["PointerId"] = 3] = "PointerId";
    PointDownEventIndex[PointDownEventIndex["X"] = 4] = "X";
    PointDownEventIndex[PointDownEventIndex["Y"] = 5] = "Y";
    PointDownEventIndex[PointDownEventIndex["EntityId"] = 6] = "EntityId";
})(PointDownEventIndex = exports.PointDownEventIndex || (exports.PointDownEventIndex = {}));
var PointMoveEventIndex;
(function (PointMoveEventIndex) {
    PointMoveEventIndex[PointMoveEventIndex["Code"] = 0] = "Code";
    PointMoveEventIndex[PointMoveEventIndex["EventFlags"] = 1] = "EventFlags";
    PointMoveEventIndex[PointMoveEventIndex["PlayerId"] = 2] = "PlayerId";
    PointMoveEventIndex[PointMoveEventIndex["PointerId"] = 3] = "PointerId";
    PointMoveEventIndex[PointMoveEventIndex["X"] = 4] = "X";
    PointMoveEventIndex[PointMoveEventIndex["Y"] = 5] = "Y";
    PointMoveEventIndex[PointMoveEventIndex["StartDeltaX"] = 6] = "StartDeltaX";
    PointMoveEventIndex[PointMoveEventIndex["StartDeltaY"] = 7] = "StartDeltaY";
    PointMoveEventIndex[PointMoveEventIndex["PrevDeltaX"] = 8] = "PrevDeltaX";
    PointMoveEventIndex[PointMoveEventIndex["PrevDeltaY"] = 9] = "PrevDeltaY";
    PointMoveEventIndex[PointMoveEventIndex["EntityId"] = 10] = "EntityId";
})(PointMoveEventIndex = exports.PointMoveEventIndex || (exports.PointMoveEventIndex = {}));
var PointUpEventIndex;
(function (PointUpEventIndex) {
    PointUpEventIndex[PointUpEventIndex["Code"] = 0] = "Code";
    PointUpEventIndex[PointUpEventIndex["EventFlags"] = 1] = "EventFlags";
    PointUpEventIndex[PointUpEventIndex["PlayerId"] = 2] = "PlayerId";
    PointUpEventIndex[PointUpEventIndex["PointerId"] = 3] = "PointerId";
    PointUpEventIndex[PointUpEventIndex["X"] = 4] = "X";
    PointUpEventIndex[PointUpEventIndex["Y"] = 5] = "Y";
    PointUpEventIndex[PointUpEventIndex["StartDeltaX"] = 6] = "StartDeltaX";
    PointUpEventIndex[PointUpEventIndex["StartDeltaY"] = 7] = "StartDeltaY";
    PointUpEventIndex[PointUpEventIndex["PrevDeltaX"] = 8] = "PrevDeltaX";
    PointUpEventIndex[PointUpEventIndex["PrevDeltaY"] = 9] = "PrevDeltaY";
    PointUpEventIndex[PointUpEventIndex["EntityId"] = 10] = "EntityId";
})(PointUpEventIndex = exports.PointUpEventIndex || (exports.PointUpEventIndex = {}));
var OperationEventIndex;
(function (OperationEventIndex) {
    OperationEventIndex[OperationEventIndex["Code"] = 0] = "Code";
    OperationEventIndex[OperationEventIndex["EventFlags"] = 1] = "EventFlags";
    OperationEventIndex[OperationEventIndex["PlayerId"] = 2] = "PlayerId";
    OperationEventIndex[OperationEventIndex["OperationCode"] = 3] = "OperationCode";
    OperationEventIndex[OperationEventIndex["Data"] = 4] = "Data";
})(OperationEventIndex = exports.OperationEventIndex || (exports.OperationEventIndex = {}));
