// @akashic-extension/akashic-timeline/lib/Timeline.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Timeline = void 0;
var Tween_1 = require("./Tween");
/**
 * タイムライン機能を提供するクラス。
 */
var Timeline = /** @class */ (function () {
    /**
     * Timelineを生成する。
     * @param scene タイムラインを実行する `Scene`
     */
    function Timeline(scene) {
        this._scene = scene;
        this._tweens = [];
        this._tweensCreateQue = [];
        this._fps = this._scene.game.fps;
        this.paused = false;
        scene.onUpdate.add(this._handler, this);
    }
    /**
     * Timelineに紐付いたTweenを生成する。
     * @param target タイムライン処理の対象にするオブジェクト
     * @param option Tweenの生成オプション。省略された場合、 {modified: target.modified, destroyed: target.destroyed} が与えられた時と同様の処理を行う。
     */
    Timeline.prototype.create = function (target, option) {
        var t = new Tween_1.Tween(target, option);
        this._tweensCreateQue.push(t);
        return t;
    };
    /**
     * Timelineに紐付いたTweenを削除する。
     * @param tween 削除するTween。
     */
    Timeline.prototype.remove = function (tween) {
        var index = this._tweens.indexOf(tween);
        var queIndex = this._tweensCreateQue.indexOf(tween);
        if (index < 0 && queIndex < 0) {
            return;
        }
        tween.cancel(false);
    };
    /**
     * Timelineに紐付いた全Tweenのアクションを完了させる。詳細は `Tween#complete()`の説明を参照。
     */
    Timeline.prototype.completeAll = function () {
        for (var i = 0; i < this._tweens.length; ++i) {
            var tween = this._tweens[i];
            if (!tween.isFinished()) {
                tween.complete();
            }
        }
        for (var i = 0; i < this._tweensCreateQue.length; ++i) {
            var tween = this._tweensCreateQue[i];
            if (!tween.isFinished()) {
                tween.complete();
            }
        }
    };
    /**
     * Timelineに紐付いた全Tweenのアクションを取り消す。詳細は `Tween#cancel()`の説明を参照。
     * @param revert ターゲットのプロパティをアクション開始前に戻すかどうか (指定しない場合は `false`)
     */
    Timeline.prototype.cancelAll = function (revert) {
        if (revert === void 0) { revert = false; }
        for (var i = 0; i < this._tweens.length; ++i) {
            var tween = this._tweens[i];
            if (!tween.isFinished()) {
                tween.cancel(revert);
            }
        }
        for (var i = 0; i < this._tweensCreateQue.length; ++i) {
            var tween = this._tweensCreateQue[i];
            if (!tween.isFinished()) {
                tween.cancel(revert);
            }
        }
    };
    /**
     * Timelineに紐付いた全Tweenの紐付けを解除する。
     */
    Timeline.prototype.clear = function () {
        this.cancelAll(false);
    };
    /**
     * このTimelineを破棄する。
     */
    Timeline.prototype.destroy = function () {
        this.clear();
        if (!this._scene.destroyed()) {
            this._scene.onUpdate.remove(this._handler, this);
        }
        this._scene = undefined;
    };
    /**
     * このTimelineが破棄済みであるかを返す。
     */
    Timeline.prototype.destroyed = function () {
        return this._scene === undefined;
    };
    Timeline.prototype._handler = function () {
        if (this.paused || this._tweens.length + this._tweensCreateQue.length === 0) {
            return;
        }
        this._tweens = this._tweens.concat(this._tweensCreateQue);
        this._tweensCreateQue = [];
        var tmp = [];
        for (var i = 0; i < this._tweens.length; ++i) {
            var tween = this._tweens[i];
            if (!tween.shouldRemove()) {
                tween._fire(1000 / this._fps);
                tmp.push(tween);
            }
        }
        this._tweens = tmp;
    };
    return Timeline;
}());
exports.Timeline = Timeline;
