// @akashic-extension/akashic-animation/lib/aps/Particle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Particle = void 0;
var Particle = /** @class */ (function () {
    function Particle(param) {
        this.elapse = 0;
        this.lifespan = param.lifespan;
        // position
        this.tx = param.tx;
        this.txMin = param.txMin;
        this.txMax = param.txMax;
        this.ty = param.ty;
        this.tyMin = param.tyMin;
        this.tyMax = param.tyMax;
        this.vx = param.vx;
        this.vy = param.vy;
        this.vMin = param.vMin;
        this.vMax = param.vMax;
        this.ax = param.ax;
        this.ay = param.ay;
        this.aMin = param.aMin;
        this.aMax = param.aMax;
        // rotation
        this.rz = param.rz;
        this.rzMin = param.rzMin;
        this.rzMax = param.rzMax;
        this.vrz = param.vrz;
        this.vrzMin = param.vrzMin;
        this.vrzMax = param.vrzMax;
        this.arz = param.arz;
        this.arzMin = param.arzMin;
        this.arzMax = param.arzMax;
        // scale
        this.sx = param.sx;
        this.sxMin = param.sxMin;
        this.sxMax = param.sxMax;
        this.vsx = param.vsx;
        this.vsxMin = param.vsxMin;
        this.vsxMax = param.vsxMax;
        this.asx = param.asx;
        this.asxMin = param.asxMin;
        this.asxMax = param.asxMax;
        this.sy = param.sy;
        this.syMin = param.syMin;
        this.syMax = param.syMax;
        this.vsy = param.vsy;
        this.vsyMin = param.vsyMin;
        this.vsyMax = param.vsyMax;
        this.asy = param.asy;
        this.asyMin = param.asyMin;
        this.asyMax = param.asyMax;
        this.sxy = param.sxy;
        this.sxyMin = param.sxyMin;
        this.sxyMax = param.sxyMax;
        this.vsxy = param.vsxy;
        this.vsxyMin = param.vsxMin;
        this.vsxyMax = param.vsxyMax;
        this.asxy = param.asxy;
        this.asxyMin = param.asxyMin;
        this.asxyMax = param.asxyMax;
        this.alpha = param.alpha;
        this.fadeInNT = param.fadeInNT;
        this.fadeOutNT = param.fadeOutNT;
    }
    return Particle;
}());
exports.Particle = Particle;
