// @akashic-extension/akashic-animation/lib/BoxVolume.js
"use strict";
var AABB = require("./AABB");
var Vector2 = require("./Vector2");
var Size2 = require("./Size2");
/**
 * BoxVolume
 *
 * アタリ判定用矩形Volume。
 *
 * 実際にアタリ判定を行うときは `BoxVolume#matrix`,`origin` そして, `size` を用いてワールド座標系での領域を求める必要があります。
 */
var BoxVolume = /** @class */ (function () {
    function BoxVolume() {
        this.origin = new Vector2();
        this.size = new Size2();
        this.aabbFirst = false;
        this.dirty = true;
    }
    BoxVolume.prototype.aabb = function () {
        if (!this._aabb || this.dirty) {
            this.dirty = false;
            var points = []; // ! 実際は CommonOffsetがpushされる
            points.push(this.matrix.multiplyPoint(this.origin));
            points.push(this.matrix.multiplyPoint(new Vector2(this.origin.x + this.size.width, this.origin.y)));
            points.push(this.matrix.multiplyPoint(new Vector2(this.origin.x, this.origin.y + this.size.height)));
            points.push(this.matrix.multiplyPoint(new Vector2(this.origin.x + this.size.width, this.origin.y + this.size.height)));
            // AABBの左上隅
            var lefttop = new Vector2(Number.MAX_VALUE, Number.MAX_VALUE);
            // AABBの右下隅
            var rightbottom = new Vector2(-Number.MAX_VALUE, -Number.MAX_VALUE);
            for (var i = 0, len = points.length; i < len; i += 1) {
                var p = points[i];
                if (p.x > rightbottom.x)
                    rightbottom.x = p.x;
                if (p.x < lefttop.x)
                    lefttop.x = p.x;
                if (p.y > rightbottom.y)
                    rightbottom.y = p.y;
                if (p.y < lefttop.y)
                    lefttop.y = p.y;
            }
            this._aabb = new AABB();
            this._aabb.origin.x = (lefttop.x + rightbottom.x) / 2;
            this._aabb.origin.y = (lefttop.y + rightbottom.y) / 2;
            this._aabb.extent.width = rightbottom.x - this._aabb.origin.x;
            this._aabb.extent.height = rightbottom.y - this._aabb.origin.y;
        }
        return this._aabb;
    };
    return BoxVolume;
}());
module.exports = BoxVolume;
